receiver/mongodbreceiver/scraper.go

```diff
@@ -0,0 +1,171 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-version"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
+)
+
+const instrumentationLibraryName = "otelcol/mongodb"
+
+type mongodbScraper struct {
+	logger       *zap.Logger
+	config       *Config
+	client       client
+	mongoVersion *version.Version
+	mb           *metadata.MetricsBuilder
+}
+
+func newMongodbScraper(logger *zap.Logger, config *Config) *mongodbScraper {
+	return &mongodbScraper{
+		logger: logger,
+		config: config,
+		mb:     metadata.NewMetricsBuilder(config.Metrics),
+	}
+}
+
+func (s *mongodbScraper) start(ctx context.Context, _ component.Host) error {
+	clientLogger := s.logger.Named("mongo-scraper")
+	c := NewClient(s.config, clientLogger)
+	s.client = c
+
+	if err := s.client.Connect(ctx); err != nil {
+		s.logger.Error("unable to connect to mongo instance", zap.Error(err))
+	}
+
+	vr, err := s.client.GetVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to get a version from the mongo instance: %w", err)
+	}
+	v, err := version.NewVersion(*vr)
+	if err != nil {
+		return fmt.Errorf("unable to parse version as semantic version %s: %w", *vr, err)
+	}
```

Comment:
Seems ok to combine these, unless there's a realistic chance that parsing the version will fail.